File: chromium-default.py
#!/usr/bin/env python3
# 6.2.0
# 2019-10-23

# Copyright (C) 2018,2019 Brandon Zorn, brandonzorn@cock.li
#
# This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License version 3
#    as published by the Free Software Foundation.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program. If not, see <https://www.gnu.org/licenses/>.

import argparse
import os
import os.path

from utils import utils


def main():
    parser = argparse.ArgumentParser()
    group1 = parser.add_argument_group('PROFILES')
    group1.add_argument('-d', '--default',
                        action='store_true',
                        help='default profile')
    group1.add_argument('-e', '--extra',
                        default='none',
                        type=str,
                        help='extra profiles, supplied arg is used as profile name')
    group1.add_argument('-s', '--sandbox',
                        action='store_true',
                        help='sandbox, all data deleted on close')
    group2 = parser.add_argument_group('BROWSER', 'which browser version to use')
    group2.add_argument('-c', '--chromium',
                        action='store_true',
                        help='chromium')
    group2.add_argument('-g', '--google-chrome',
                        default='unstable',
                        type=str,
                        help='chrome, valid are [unstable,beta,release]')
    args = parser.parse_args()

    utils.is_not_root()

    if args.chromium:
        cmd = 'chromium'
    elif args.google_chrome == 'unstable':
        cmd = 'google-chrome-unstable'
    elif args.google_chrome == 'beta':
        cmd = 'google-chrome-beta'
    else:
        cmd = 'google-chrome'

    profile = os.path.join(os.environ['HOME'], '.config/chrome')

    if args.extra is not 'none':
        profile_append = f'{cmd}-{args.extra}'
    else:
        # script/symlinks must be named chromium-<profile name> to work
        # or this can be changed
        profile_append = f'{cmd}-{utils.get_script_name()[9:]}'

    if args.sandbox:
        from tempfile import TemporaryDirectory
        with TemporaryDirectory() as tmpdir:
            cmd += f' --user-data-dir={tmpdir}'
            utils.run_cmd(cmd)
    else:
        profile += f'/{profile_append}'
        cmd += f' --user-data-dir={profile}'
        utils.run_cmd(cmd)


if __name__ == "__main__":
    main()

File: cxx/chromium-default.cxx
/**
 * Copyright (C) 2024 Brandon Zorn <brandonzorn@cock.li>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <string>

#include <format>

#include <array>

#include <source_location>

#include <glibmm.h>

#include <CLI/CLI.hpp>

#include <ztd/ztd.hxx>
#include <ztd/ztd_logger.hxx>

#include "lib/env.hxx"
#include "lib/commandline.hxx"
#include "lib/user-dirs.hxx"

const auto package = package_data{
    std::source_location::current().file_name(),
    "2024-01-04",
    "9.0.0",
};

#define LAUNCH_ASYNC

int
main(int argc, char** argv)
{
    CLI::App app{"Chrome multiple profiles launcher"};

    const std::array<std::string, 3> chrome_bins = {
        "google-chrome-stable",
        "google-chrome-unstable",
        "google-chrome-beta",
    };
    std::string chrome_bin_name = chrome_bins[0];
    app.add_option("-c,--chrome", chrome_bin_name, "chrome binary to use")
        ->expected(1)
        ->check(CLI::IsMember(chrome_bins));

    auto opt = commandline_opt_data::create(package);

    setup_common_commandline(app, opt, false);

    CLI11_PARSE(app, argc, argv);

    env::check_running_user(env::only_run_as::user);

    // need to know display server type for ozone to work
    const std::string display_server = env::is_wayland() ? "wayland" : "x11";

    // symlinks must be named 'chromium-<profile>' to work.
    // e.g. ln -s chromium-default chromium-<profile>
    const std::filesystem::path launch_name = argv[0];
    const std::string profile_name =
        ztd::removeprefix(launch_name.filename().string(), "chromium-");

    // Get chrome bin name from file path
    std::string bin_name = chrome_bin_name;
    if (bin_name.contains('/'))
    {
        bin_name = ztd::rpartition(chrome_bin_name, "/")[2];
    }

    // Build config path
    const std::string chrome_profile = std::format("{}-{}", bin_name, profile_name);
    const std::string profile_path =
        Glib::build_filename(user::config_dir(), "chrome", chrome_profile);

    // ztd::logger::info("argv[0]        = {}", argv[0]);
    // ztd::logger::info("profile_name   = {}", profile_name);
    // ztd::logger::info("chrome_profile = {}", chrome_profile);
    // ztd::logger::info("profile_path   = {}", profile_path);

    const std::string command = std::format("{} "
                                            "--user-data-dir={} "
                                            "--ozone-platform={} "
                                            "--no-default-browser-check ",
                                            chrome_bin_name,
                                            profile_path,
                                            display_server);

#if defined(LAUNCH_ASYNC)
    Glib::spawn_command_line_async(command);
    std::exit(EXIT_SUCCESS);
#else
    i32 exit_status;
    Glib::spawn_command_line_sync(command, nullptr, nullptr, &exit_status);
    std::exit(exit_status);
#endif
}
